### ISSUE TITLE
CPU Training error:  smooth_L1_loss_layer Not Implemented Yet

### ISSUE THREAD

**Issue by sunshineatnoon** [rbgirshick/fast-rcnn#53]
I modified train_net.py to use cpu for training. I also disabled MATLAB and use the pre-computed selective search bounding box for VOC2007. However, I got this error during training:
```
Solving...
F0906 14:48:34.166416 13116 smooth_L1_loss_layer.cpp:39] Not Implemented Yet
*** Check failure stack trace: ***
```
I looked into fast-rcnn/caffe-fast-rcnn/src/caffe/layers/smooth_L1_loss_layer.cpp and found following code:
```
template <typename Dtype>
void SmoothL1LossLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
    const vector<Blob<Dtype>*>& top) {
  NOT_IMPLEMENTED;
}

template <typename Dtype>
void SmoothL1LossLayer<Dtype>::Backward_cpu(const vector<Blob<Dtype>*>& top,
    const vector<bool>& propagate_down, const vector<Blob<Dtype>*>& bottom) {
  NOT_IMPLEMENTED;
}
```
So I guess this layer is not implemented by this version of caffe. But SmoothL1LossLayer is implemented for GPU. I also looked up in the newest version of caffe, but there is even no such file called smooth_L1_loss_layer.cpp.
Does this mean I can't use CPU to train network? Or do I have to use another kind of loss layer?

**Comment by kaifeichen** [comment]
+1

**Comment by haimansx** [comment]
@sunshineatnoon @kaifeichen @rbgirshick   I got the same problem. Did you solve this problem ???

**Comment by sunshineatnoon** [comment]
@haimansx No, I have not solved the problem, I think the only way to solve the problem is by implementing the two functions above in c++

**Comment by kaifeichen** [comment]
Hi @haimansx,  I think @sunshineatnoon is right. I just used the GPU version and it worked.

**Comment by tanjoreg** [comment]
Hi @sunshineatnoon I have followed your blog to train Fast-RCNN my own dataset and i have stumbled upon this error. I am using CPU Mode. Were you able to solve this issue?

@kaifeichen Does that mean Fast-RCNN is not implemented for CPU mode? Any updates on this would help me.

**Comment by sunshineatnoon** [comment]
@tanjoreg No, I haven't solved this issue. Fast Rcnn can only be trained on GPU as far as I know. But it can be tested on CPU.

**Comment by tanjoreg** [comment]
@sunshineatnoon Thank you very much,for the swift reply. Your blog helped me a lot.
I have tested the imagenet demo on CPU and it works. I was trying to train my own dataset, which resulted in the error. Does that mean I have hit the wall with CPU mode?

**Comment by sunshineatnoon** [comment]
As I mentioned above, I think the way to solve this issue is to implement the two functions above in c++.

**Comment by hnflee** [comment]
+1

**Comment by athulase** [comment]
@sunshineatnoon: Were you able to train R-CNN in cpu? or is it still in the status of not running training on cpu mode?

**Comment by HamdiHamed1992** [comment]
I am facing the same problem!  
Any help?  
If someone implemented those 2 functions in cpp, can he post the code?

**Comment by PeterJiangwy** [comment]
@HamdiHamed1992 , Do you solve this problem. There is someone write these 2 functions, but it still not work.

**Comment by PeterJiangwy** [comment]
@HamdiHamed1992 , Do you solve this problem. There is someone write these 2 functions, but it still not work.

**Comment by HamdiHamed1992** [comment]
@PeterJiangwy sorry no. 
I am using a CPU to train CNN and interested to segmentation not object localisation. Therefore i did not focus alot with RCNN, Fast RCNN and Faster RCNN.
I can suggest you to use other network architectures and change just the final classification layers to your wanted task. 
Choose some already existing models that can run on CPU... (Segnet 4 example, based on VGG16 Model arch).
If you need anything, feel free to ask.

**Comment by PeterJiangwy** [comment]
@ HamdiHamed1992, Thank you.
I’m newer to this scope. Now I’m trying to do something about object detection. Do you have any suggestions?
Thank you very much for your help.
